### ISSUE TITLE
Huge Plex library (>6000 movies) crashes Kodi on initial sync

### ISSUE THREAD

**Issue by danjames92** [croneter/PlexKodiConnect#531]
Crash on movie item number 6691 (Home Alone 2 according to onscreen GUI).

Log 1 (log 2 to follow)

https://drive.google.com/open?id=1Wcsvmb4_bos0Ro5vVfGEnH53_vR726Nv

Log 2 @croneter this one on 6616 (The Guardian according to onscreen GUI).

https://drive.google.com/open?id=1R5EMSCHIo4zpQGuFfpes_FtOx790lOvz

**Comment by croneter** [comment]
There is nothing at all in the log indicating an issue with neither PKC nor Kodi, unfortunately.

I must tell you that Kodi 17.6 is not the most stable piece of software. I myself get crashes from time to time. This should improve with Kodi 18, once it's out of beta phase. 

I'll close this issue as there is nothing I can do, unfortunately.

**Comment by danjames92** [comment]
Shame as even after a complete fresh install it seems broken on a Pi 3. I take it you don't have one to test with? 

That crash dump I PM'D you was from the Pi 3 by the way.

**Comment by danjames92** [comment]
I doubt I'll get anywhere with them and it has worked before without issue via LibreELEC even with 8.2.5 so I have no idea whats causing it. It seems like the load of the initial sync or maybe my huge library size just freezes Kodi dead. I chose the Pi 3 cause it's the only device that does full resolution 3D MVC playback. Something i can only get in Kodi. :)

**Comment by croneter** [comment]
The testing would be solely to check whether another Kodi, outside the LibreElec world, also crashes. 

PKC resumes sync after a crash - where it left off. Does your entire library sync after a couple of crashes?

**Comment by danjames92** [comment]
That's the thing. It starts from scratch. There's no resuming and it beginnings the 7K ish sync from the beginning again.

Would you suggest any particular flavour of Kodi?

OpenELEC maybe?

**Comment by croneter** [comment]
Weird... That indicates that the Kodi database gets reset.

Try pulling the plug after 5k each time

**Comment by croneter** [comment]
If you don't mind me having access to your library, could you please temporarily add me as a friend (username croneter)? I could thus try to reproduce the issue. My library is way too small right now :wink:

**Comment by danjames92** [comment]
Invited.

**Comment by croneter** [comment]
Please try the new BETA version 2.3.11 and report back

**Comment by danjames92** [comment]
Will give it a go and report back! Thank you!

**Comment by danjames92** [comment]
@croneter not fixed. froze at 6499 items.

**Comment by danjames92** [comment]
Can you confirm it happens on your Pi 3 with Librelec as well?

On 31 Aug 2018, at 21:24, croneter <notifications@github.com<mailto:notifications@github.com>> wrote:


I just synced all your 8000+ movies plus the rest to Kodi 17.6 on Windows 10. Please take this issue to the LibreElec guys...

[image]<https://user-images.githubusercontent.com/4637940/44934304-a3bdd380-ad6c-11e8-9a7e-63d31782cc04.png>

**Comment by danjames92** [comment]
@croneter when you're back would you mind confirming if it crashes for you? I'll then reach out the LibreELEC guys after. 👍

**Comment by danjames92** [comment]
Done.

**Comment by danjames92** [comment]
Wasn't caching images I think with a 16GB card. I'll have to confirm. 

Just a quick FYI, "vMovies -Remux" is the start of me rebuilding my libraries using /data instead of /mnt/unionfs as a root path.

**Comment by croneter** [comment]
I can confirm that my LibreElec crashes around movie number ~6100. Sync with Kodi 17.6 on Windows works, though. Hence: Nothing I can do at the moment from the PKC side, then...

**Comment by danjames92** [comment]
How many sync threads were you using? I found sometimes it got further using 1.

**Comment by croneter** [comment]
Nope the number of threads seemed random. And definitely no - Kodi on Windows works. I will not implement a work-around for a single platform that is bound to be updated very soon anyway, sorry. Besides the number of writes per second does not seem to be the issue, otherwise the library scan would crash sooner

**Comment by milnash** [comment]
Same happens with Nvidia Shield on a huge library. My crash seems to be because of RAM running out. And it is not my movies, 7000+, it is my TV Show episodes, 75 000+. Using Kodi 17.6. 

Work around: 
1. I manually exit Kodi before RAM running out and causing the crash.
2. Start Kodi again.
3. Wait couple of minutes and then the sync continues.

Note: If I let Kodi crash, the sync starts all over again.

**Comment by milnash** [comment]
> Same happens with Nvidia Shield on a huge library. My crash seems to be because of RAM running out. And it is not my movies, 7000+, it is my TV Show episodes, 75 000+. Using Kodi 17.6.
> 
> # Work around:
> 1. I manually exit Kodi before RAM running out and causing the crash.
> 2. Start Kodi again.
> 3. Wait couple of minutes and then the sync continues.
> 
> Note: If I let Kodi crash, the sync starts from scratch again.

I was wrong. My work around does not work. Have been trying for 2 days to get my sync working...
Hope someone figures this out.

**Comment by croneter** [comment]
@milnash Could you please share a debug log file? This bug will be very hard to pin down

**Comment by croneter** [comment]
@milnash Does this also happen if you PKC and Kodi on e.g. a Windows machine?

**Comment by milnash** [comment]
> @milnash Does this also happen if you PKC and Kodi on e.g. a Windows machine?

No. Only with my Nvidia Shield and I guess it's because my Windows machine has 16GB RAM. 
Is it possible to add something in the code to free up the memory if it reaches below let's say 300MB? Free memory is visible in Kodi. I'm no coder as you can guess :)

**Comment by milnash** [comment]
Yes, the RAM decreases on the Shield during the sync and then crashes when it reaches 180 MB free. 
I can try with a Windows machine - but I guess you also have a Windows machine? I can share my Plex server with you so you can test and play movies or shows if you want :)

**Comment by croneter** [comment]
Excited to hear back from you - I think I got the buggyer :laughing:

**Comment by danjames92** [comment]
Oh this is exciting! I will give it a go again. :)

**Comment by croneter** [comment]
Hmm, does not seem to be quite fixed yet :disappointed: 

I think we're looking at some kind of Python "arena fragmentation bug" caused by xml.etree.ElementTree, see https://bugs.python.org/issue11849

**Comment by croneter** [comment]
Also see here: https://benbernardblog.com/tracking-down-a-freaky-python-memory-leak/

**Comment by croneter** [comment]
It's much better, but my Pi 3 still crashes

**Comment by danjames92** [comment]
Yeah you're right. I get crashes with 8.2.5 too. Was ok with CoreELEC 8.95.3 but that's not what you're looking at here.

**Comment by jmurph116** [comment]
Just like to point out that I still haven't been able to get PKC to sync my library on any of my devices either. Tried using kodi 17.6 and Kodi 18 beta 3 on multiple devices.  I too, have an incredibly large library. Kodi crashes around the 8k mark of movies and then the sync will start from scratch upon reboot.

**Comment by croneter** [comment]
Even now, though it's called 2.4.10 ;-)

Looking solved on my Raspberry Pi 3 :laughing:

**Comment by croneter** [comment]
@danjames92
Your library is so huge I now need to rewire the sync process... :-). Stay tuned

**Comment by danjames92** [comment]
I try my hardest to make your project fun to work on... not inadvertent at all ;) Yep, I was just about to say it's still crashing with 2.4.10 on my Odroid C2, I see the ram usage drop and drop and drop until about 200MB free where it seems to freeze.

:)

**Comment by croneter** [comment]
I should be able to push a new beta version in the next few days. This is a GIANT rewrite....

**Comment by danjames92** [comment]
@croneter eta wen son :)

**Comment by croneter** [comment]
Please try the new PKC **beta** version 2.5.0. Problem solved?

**Comment by danjames92** [comment]
Will wait for `2.5.10` for V18 Leia login fix. 👍

**Comment by croneter** [comment]
Huh? Try 2.5.1 now. Login Dialog as well as Choose Plex Home User dialogs should be fixed, if that's what you mean

**Comment by danjames92** [comment]
Only 2.5.0 was out when I checked.

Checking out 2.5.1 now... Will report back.

On 2 Dec 2018, at 17:02, croneter <notifications@github.com<mailto:notifications@github.com>> wrote:


Huh? Try 2.5.1 now. Login Dialog as well as Choose Plex Home User dialogs should be fixed, if that's what you mean

**Comment by milnash** [comment]
The initial sync seems slower but RAM usage seems very stable. 20K episodes of 100K in 50 minutes. Hopefully Kodi can handle this large library after sync 😂

Will report back soon. Thank you for all your effort and work on this!! 👌👏🙏

**Comment by croneter** [comment]
Hmmm, still need to rewire the sync a bit, it seems. The sync flies for my library

**Comment by danjames92** [comment]
I was testing on LibreELEC 8.9.000, the latest Leia RC1.

On 3 Dec 2018, at 07:04, croneter <notifications@github.com<mailto:notifications@github.com>> wrote:


Hmmm, still need to rewire the sync a bit, it seems. The sync flies for my library

**Comment by danjames92** [comment]
Will give it a go tomorrow. 👍

**Comment by milnash** [comment]
Works awesome now.
Nvidia Shield and 15k+ movies. 

Thanks!

**Comment by croneter** [comment]
@milnash 
Nice! Is the sync process faster, too?

**Comment by danjames92** [comment]
@milnash any idea of your library size so I've got something to compare against? 

@croneter does this new beta still require a full initial sync on start up every time you reboot?

**Comment by milnash** [comment]
@danjames92 Just did 9800 movies and 2600 TV shows. Much faster now than any other version I've tried. 

Note: it might look like it's doing 1 movie / second but then suddenly it jumps up couple of hundreds/thousands. So just let it finish :) 

And no it's not doing initial sync on reboot any more.

**Comment by danjames92** [comment]
@milnash 2600 episodes or 2600 shows?

Just clean installed LibreELEC (Leia) v8.90.009 ALPHA on my Odroid C2 and kicked off my initial sync. *fingers crossed*

**Comment by milnash** [comment]
2600 shows. Just tried version 2.5.5 on another box, and same behaviour as you regarding double reboot. Seems OK noe though :)

**Comment by croneter** [comment]
Double reboot is necessary cause PKC is messing with the Kodi DB a bit more heavily

**Comment by Zixim** [comment]
this isn't in stable yet?

**Comment by MacQwerty** [comment]
@Zixim, not yet. However, I assume @croneter is going to update the stable version in the very close future as the latest beta versions seem to be quite stable

**Comment by Zixim** [comment]
I installed latest from Beta repo...
syncing 11k movies & 78k TV episodes.
Can't take my eyes off the progress indicator.

**Comment by Zixim** [comment]
sadpanda
😒 it crashed _after_ the sync.
On reboot it started the sync from scratch.

**Comment by croneter** [comment]
Darn - it worked for the others with huge libraries 😢 . Are you using a stable Kodi version like Kodi 17.6?

Please share your log (delete your Plex tokens) with the crash. See https://github.com/croneter/PlexKodiConnect/wiki/How-to-Report-A-Bug

**Comment by Zixim** [comment]
it's a Pi 3 B running osmc, Kodi 17.6
The thing is now totally unresponsive, including SSH.
I'll try again from scratch, and grab a log.

I'm none too bothered, this is an experimental setup. I'm testing my options to move away from Rasplex.

**Comment by Zixim** [comment]
hth : https://paste.ubuntu.com/p/3kJMrMzK6M/

**Comment by Zixim** [comment]
I have started over from scratch since last post.
On OSMC the ram decreased to about 48.000KB, then crash. Every time.

installed LibreElec,
turned on debug logging,
installed PKC 2.5.22,
configured PKC,
it is now syncing...not finished yet.
The memory hasn't dropped below 500.000KB yet.

@croneter (how) can i send this log directly to you, in case it crashes again ?

**Comment by croneter** [comment]
Sound like a memory leak for OSMC for now...

Just drop the log here as a file, remember to remove to tokens!

**Comment by Zixim** [comment]
It chewed through the sync in OSMC...but afterwards it crashed, and seemed to start from 0 on reboot, and crashed, and ...

Log is 8megs now...64% synced. 😨

**Comment by croneter** [comment]
Hmm, then OSMC just could not handle your giant Kodi sqlite database 😆 

Fingers crossed for LibreElec!

**Comment by Zixim** [comment]
it synced, crashed only once...
But browsing in the media library is unfeasably slow.
I guess you must be right, this is too big for the humble pi.
Does any one see a solution, or should I just move along and try something else?

I logged then entire sync process, including where it crashed, but the file is 13MB.
[kodi.zip](https://github.com/croneter/PlexKodiConnect/files/2804709/kodi.zip)

**Comment by croneter** [comment]
Closing this for now. 

@Zixim Jupp I think you're giant DB is too much for the humble Pi to handle.

**Comment by Zixim** [comment]
for posterity : 
The library size is about 95.000 items, give or take a couple of thousands.

Many thanks to @croneter for digging into massive logs, to no avail.

If/when I upgrade the hardware, i'm sure to revisit PKC.
